File: 2316. Count Unreachable Pairs of Nodes in an Undirected Graph/Solution.java
class Solution {
    public long countPairs(int n, int[][] edges) {
        UFDS uf = new UFDS(n);
        for (int[] e : edges) uf.union(e[0], e[1]);
        int[] reps = uf.getReps();
        long res = 0;
        for (int r : reps) {
            n -= r;
            res += 1L * n * r;
        }
        return res;
    }
}

class UFDS {
    int[] parent, rank, size;
    int disjoint, n;

    public UFDS(int n) {
        parent = new int[n];
        rank = new int[n];
        size = new int[n];
        disjoint = n;
        this.n = n;
        for (int i = 0; i < n; i++) {
            parent[i] = i;
            size[i] = 1;
        }
    }

    public int find(int x) {
        if (parent[x] != x) parent[x] = find(parent[x]);
        return parent[x];
    }

    public boolean same(int x, int y) {
        return find(x) == find(y);
    }

    public void union(int x, int y) {
        if (same(x, y)) return;
        int a = find(x), b = find(y);
        if (rank[a] > rank[b]) {
            parent[b] = a;
            size[a] += size[b];
        } else {
            parent[a] = b;
            size[b] += size[a];
            if (rank[a] == rank[b]) rank[b]++;
        }
        disjoint--;
    }

    public int[] getReps() {
        int[] reps = new int[disjoint];
        int pointer = 0;
        for (int i = 0; i < n; i++) {
            if (parent[i] == i) {
                reps[pointer] = size[i];
                pointer++;
            }
        }
        return reps;
    }
}

File: 2316. Count Unreachable Pairs of Nodes in an Undirected Graph/Solution.py
class Solution:
    def countPairs(self, n: int, edges: List[List[int]]) -> int:
        uf = UFDS(n)
        for x, y in edges: uf.union(x, y)
        reps = uf.getReps()
        res = 0
        for r in reps:
            n -= r
            res += n * r
        return res

class UFDS:
    def __init__(self, n):
        self.parent = [i for i in range(n)]
        self.rank = [0 for i in range(n)]
        self.size = [1 for i in range(n)]
        self.disjoint = n
        self.n = n
    
    def find(self, x):
        if self.parent[x] != x: self.parent[x] = self.find(self.parent[x])
        return self.parent[x]
    
    def same(self, x, y):
        return self.find(x) == self.find(y)

    def union(self, x, y):
        if self.same(x, y): return
        a, b = self.find(x), self.find(y)
        if self.rank[a] > self.rank[b]:
            self.parent[b] = a
            self.size[a] += self.size[b]
        else:
            self.parent[a] = b
            self.size[b] += self.size[a]
            if self.rank[a] == self.rank[b]: self.rank[b] += 1
        self.disjoint -= 1
    
    def getReps(self):
        return [self.size[i] for i in range(self.n) if self.parent[i] == i]